File: haskell/collatz-conjecture/src/CollatzConjecture.hs
module CollatzConjecture (collatz) where

collatz :: Integer -> Maybe Integer
collatz n
    | n <= 0 = Nothing
    | otherwise = Just $ findCollatz n
    where
        next i = if even i then i `div` 2 else i * 3 + 1
        findCollatz i = toInteger $ length $ takeWhile (\x -> x /= 1) $ iterate next n

File: scala/collatz-conjecture/src/main/scala/CollatzConjecture.scala
object CollatzConjecture {

  def steps(n: Int): Option[Int] = {
    if (n <= 0)
      None
    else
      Option(Stream.iterate(n)(next).takeWhile(i => i != 1).length)
  }

  def next(n: Int): Int = {
    if (n % 2 == 0)
      n / 2
    else
      n * 3 + 1
  }
}
